### ISSUE TITLE
[IoT Broker QA] How to use NGSI association feature.

### ISSUE THREAD

**Issue by onlyindreams** [Aeronbroker/Aeron#8]
I'm trying to use a NGSI-Association feature which FIWARE IoT Broker provides.

https://forge.fiware.org/plugins/mediawiki/wiki/fiware/index.php/NGSI_association

I have registerd following two contexts to IoT Discovery(NEConfman).

1). farm context
```
(curl localhost:8060/ngsi9/registerContext -s -S \
    --header  'Content-Type: application/xml' --header 'Accept: application/xml' \
    -d @-| xmllint --format - ) <<EOF
<?xml version="1.0" encoding="UTF-8"?>
<registerContextRequest>
<contextRegistrationList>
    <contextRegistration>
        <entityIdList>
            <entityId type="farm" isPattern="false">
                <id>farm1</id>
            </entityId>
        </entityIdList>
        <providingApplication></providingApplication>
    </contextRegistration>
</contextRegistrationList>
<duration>P1Y</duration>
<registrationId></registrationId>
</registerContextRequest>
EOF
```

2). thermo context
```
(curl localhost:8060/ngsi9/registerContext -s -S \
    --header  'Content-Type: application/xml' --header 'Accept: application/xml' \
    -d @-| xmllint --format - ) <<EOF
<?xml version="1.0" encoding="UTF-8"?>
<registerContextRequest>
<contextRegistrationList>
    <contextRegistration>
        <entityIdList>
            <entityId type="thermometer" isPattern="false">
                <id>thermo1</id>
            </entityId>
        </entityIdList>
        <contextRegistrationAttributeList>
        <contextRegistrationAttribute>
            <name>temperature</name>
            <type>float</type>
            <isDomain>false</isDomain>
            <metadata>
                        <contextMetadata>
                            <name>unit</name>
                            <type>string</type>
                        </contextMetadata>
            </metadata>
        </contextRegistrationAttribute>
        <contextRegistrationAttribute>
            <name>humidity</name>
            <type>float</type>
            <isDomain>false</isDomain>
            <metadata>
                        <contextMetadata>
                            <name>unit</name>
                            <type>string</type>
                        </contextMetadata>
            </metadata>
        </contextRegistrationAttribute>
        <contextRegistrationAttribute>
            <name>measurement_time</name>
            <type>timestamp</type>
            <isDomain>false</isDomain>
        </contextRegistrationAttribute>
        </contextRegistrationAttributeList>
        <providingApplication>http://localhost:1027/ngsi10</providingApplication>
    </contextRegistration>
</contextRegistrationList>
<duration>P1Y</duration>
<registrationId></registrationId>
</registerContextRequest>
EOF
```

A providing application(localhost:1027) has this context data(thermo) values.

And I have registerd a association context related to the farm context and the thermo context.
```
(curl localhost:8060/ngsi9/registerContext -s -S \
    --header  'Content-Type: application/xml' --header 'Accept: application/xml' \
    -d @-) <<EOF
<?xml version="1.0" encoding="UTF-8"?>
<registerContextRequest>
<contextRegistrationList>
        <contextRegistration>
            <registrationMetadata>
                <contextMetadata>
                    <name>farm1-association1</name>
                    <type>Association</type>
                    <value>
                        <entityAssociation>
                            <sourceEntityId type="thermometer" isPattern="false">
                                <id>thermo1</id>
                            </sourceEntityId>
                            <targetEntityId type="farm" isPattern="false">
                                <id>farm1</id>
                            </targetEntityId>
                        </entityAssociation>
                    </value>
                </contextMetadata>
            </registrationMetadata>
            <providingApplication></providingApplication>
        </contextRegistration>
    </contextRegistrationList>
    <duration>P1Y</duration>
    <registrationId></registrationId>
</registerContextRequest>
EOF
```

These context registrations has successfully done but a query has failed.
```
# curl http://localhost:8060/ngsi10/contextEntities/farm1
<!DOCTYPE html>
<html lang="en">
    <head>
                <meta charset="UTF-8" />
        <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1"> 
                <meta name="viewport" content="width=device-width, initial-scale=1.0"> 
        <title>IoT Broker Admin Login</title>      
        <link rel="shortcut icon" href="../favicon.ico"> 
        <link rel="stylesheet" type="text/css" href="/resources/logins/style.css" />
                <!--[if lte IE 7]><style>.main{display:none;} .support-note .note-ie{display:block;}</style><![endif]-->
                <style>

                </style>
    </head>
    <body>
        <div class="container">

                <header>

                                <h1>NEC <strong>IoT Broker GE</strong></h1>
                                <div class="support-note">
                                        <span class="note-ie">Sorry, only modern browsers.</span>
                                </div>
                                <h1><strong><font size="20">Back Home!</font></strong></h1>

                        </header>
        </div>
    </body>
</html> 
```

IoT Broker has output following error.
```
2016-08-16 10:05:29 ERROR [spring]:843 - Servlet.service() for servlet spring threw exception
java.lang.NullPointerException
        at eu.neclab.iotplatform.iotbroker.core.IotBrokerCore.queryContext(IotBrokerCore.java:675)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
        at org.springframework.osgi.service.importer.support.internal.aop.ServiceInvoker.doInvoke(ServiceInvoker.java:58)
        at org.springframework.osgi.service.importer.support.internal.aop.ServiceInvoker.invoke(ServiceInvoker.java:62)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
        at org.springframework.aop.support.DelegatingIntroductionInterceptor.doProceed(DelegatingIntroductionInterceptor.java:132)
        at org.springframework.aop.support.DelegatingIntroductionInterceptor.invoke(DelegatingIntroductionInterceptor.java:120)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
        at org.springframework.osgi.service.util.internal.aop.ServiceTCCLInterceptor.invokeUnprivileged(ServiceTCCLInterceptor.java:71)
        at org.springframework.osgi.service.util.internal.aop.ServiceTCCLInterceptor.invoke(ServiceTCCLInterceptor.java:54)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
        at org.springframework.osgi.service.importer.support.LocalBundleContextAdvice.invoke(LocalBundleContextAdvice.java:59)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
        at org.springframework.aop.support.DelegatingIntroductionInterceptor.doProceed(DelegatingIntroductionInterceptor.java:132)
        at org.springframework.aop.support.DelegatingIntroductionInterceptor.invoke(DelegatingIntroductionInterceptor.java:120)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204)
        at com.sun.proxy.$Proxy21.queryContext(Unknown Source)
        at eu.neclab.iotplatform.iotbroker.restcontroller.RestProviderController.simpleQueryIdGet(RestProviderController.java:610)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.web.method.support.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:219)
        at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:132)
        at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:104)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandleMethod(RequestMappingHandlerAdapter.java:745)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:686)
        at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:80)
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:925)
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:856)
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:936)
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:827)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:690)
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:812)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:803)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:290)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:197)
        at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:171)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
        at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:118)
        at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:84)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
        at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:113)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
        at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:103)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
        at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:113)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
        at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:54)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
        at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:45)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
        at org.springframework.security.web.authentication.www.BasicAuthenticationFilter.doFilter(BasicAuthenticationFilter.java:150)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
        at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:183)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
        at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:105)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
        at org.springframework.security.web.session.ConcurrentSessionFilter.doFilter(ConcurrentSessionFilter.java:125)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
        at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:87)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
        at org.springframework.security.web.access.channel.ChannelProcessingFilter.doFilter(ChannelProcessingFilter.java:144)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:192)
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:160)
        at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:346)
        at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:259)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:233)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:191)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:128)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:102)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:109)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:286)
        at org.apache.coyote.http11.Http11Processor.process(Http11Processor.java:845)
        at org.apache.coyote.http11.Http11Protocol$Http11ConnectionHandler.process(Http11Protocol.java:583)
        at org.apache.tomcat.util.net.JIoEndpoint$Worker.run(JIoEndpoint.java:447)
        at java.lang.Thread.run(Thread.java:745)
```

I suspect my understanding of the association feature.
Could you tell me how to use a NGSI-association feature in this case?

Best Regards,
Naoya

**Comment by Aeronbroker** [comment]
Dear Naoya,

the issue relies on the fact that you are providing an empty providingApplication field in the first and the third messages which is not allowed in a correct NGSI9 message. I have updated the XML schema in order to specify this restriction and avoid such incorrect registerContext to be stored in the IoT Discovery.

Please try to make again the same steps but with adding the providingApplication url. Furthermore try to make a git pull in order to update the xml schema files (in order to avoid this kind of issue in the future).

Best Regards
Flavio

**Comment by onlyindreams** [comment]
Thank you for your response.

I understood that Association contexts and All of Entity Contexts need "providingApplication url".

Then, I have Re-tried but it doesn't work well..

- farm context
```
(curl localhost:8060/ngsi9/registerContext -s -S \
    --header  'Content-Type: application/xml' --header 'Accept: application/xml' \
    -d @-| xmllint --format - ) <<EOF
<?xml version="1.0" encoding="UTF-8"?>
<registerContextRequest>
<contextRegistrationList>
    <contextRegistration>
        <entityIdList>
            <entityId type="farm" isPattern="false">
                <id>farm1</id>
            </entityId>
        </entityIdList>
        <contextRegistrationAttributeList>
        <contextRegistrationAttribute>
            <name>farm1_temp</name>
            <type>float</type>
            <isDomain>false</isDomain>
        </contextRegistrationAttribute>
        </contextRegistrationAttributeList>
        <providingApplication>http://localhost:1027/ngsi10</providingApplication>
    </contextRegistration>
</contextRegistrationList>
<duration>P1Y</duration>
<registrationId></registrationId>
</registerContextRequest>
EOF
```

- thermometer context
```
(curl localhost:8060/ngsi9/registerContext -s -S \
    --header  'Content-Type: application/xml' --header 'Accept: application/xml' \
    -d @-| xmllint --format - ) <<EOF
<?xml version="1.0" encoding="UTF-8"?>
<registerContextRequest>
<contextRegistrationList>
    <contextRegistration>
        <entityIdList>
            <entityId type="thermometer" isPattern="false">
                <id>thermo1</id>
            </entityId>
        </entityIdList>
        <contextRegistrationAttributeList>
        <contextRegistrationAttribute>
            <name>temperature</name>
            <type>float</type>
            <isDomain>false</isDomain>
            <metadata>
                        <contextMetadata>
                            <name>unit</name>
                            <type>string</type>
                        </contextMetadata>
            </metadata>
        </contextRegistrationAttribute>
        <contextRegistrationAttribute>
            <name>humidity</name>
            <type>float</type>
            <isDomain>false</isDomain>
            <metadata>
                        <contextMetadata>
                            <name>unit</name>
                            <type>string</type>
                        </contextMetadata>
            </metadata>
        </contextRegistrationAttribute>
        <contextRegistrationAttribute>
            <name>measurement_time</name>
            <type>timestamp</type>
            <isDomain>false</isDomain>
        </contextRegistrationAttribute>
        </contextRegistrationAttributeList>
        <providingApplication>http://localhost:1027/ngsi10</providingApplication>
    </contextRegistration>
</contextRegistrationList>
<duration>P1Y</duration>
<registrationId></registrationId>
</registerContextRequest>
EOF
```

- AssociationContext
```
(curl localhost:8060/ngsi9/registerContext -s -S \
    --header  'Content-Type: application/xml' --header 'Accept: application/xml' \
    -d @-) <<EOF
<?xml version="1.0" encoding="UTF-8"?>
<registerContextRequest>
<contextRegistrationList>
        <contextRegistration>
            <registrationMetadata>
                <contextMetadata>
                    <name>farm1-association1</name>
                    <type>Association</type>
                    <value>
                        <entityAssociation>
                            <sourceEntityId type="thermometer" isPattern="false">
                                <id>thermo1</id>
                            </sourceEntityId>
                            <targetEntityId type="farm" isPattern="false">
                                <id>farm1</id>
                            </targetEntityId>
                        </entityAssociation>
			<AttributeAssociationList>
				<AttributeAssociation>
        			<sourceAttribute>temperature</sourceAttribute>
       				<targetAttribute>farm1_temp</targetAttribute>
				</AttributeAssociation>
			</AttributeAssociationList>
                    </value>
                </contextMetadata>
            </registrationMetadata>
            <providingApplication>http://localhost:1027/ngsi10</providingApplication>
        </contextRegistration>
    </contextRegistrationList>
    <duration>P1Y</duration>
    <registrationId></registrationId>
</registerContextRequest>
EOF
```

Localhost:1027 can provide the following datas ( farm1 and thermometer both).
```
(curl localhost:1027/v1/updateContext -s -S --header 'Content-Type: application/json' \
    --header 'Accept: application/json' -d @- | python -mjson.tool) <<EOF
{
    "contextElements": [
        {
            "type": "farm",
            "isPattern": "false",
            "id": "farm1",
            "attributes": [{"name":"farm1_temp","type":"float","value":"0"}]
        },
        {
            "type": "thermometer",
            "isPattern": "false",
            "id": "thermo1",
            "attributes": [
                {
                    "name": "temperature",
                    "type": "float",
                    "value": "20.00",
                    "metadatas": [
                       {
                            "name" : "unit",
                            "type" : "string",
                            "value" : "celsius"
                       }
                    ]
                },
                {
                    "name": "humidity",
                    "type": "float",
                    "value": "30.00",
                    "metadatas": [
                       {
                            "name" : "unit",
                            "type" : "string",
                            "value" : "%"
                       }
                    ]
                }
            ]
        }
    ],
    "updateAction": "APPEND"
} 
EOF
```

Then I issued queryContext to IoT Broker but a response is...
```
curl http://localhost:8060/ngsi10/contextEntities/farm1 | xmllint --format -
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
110   552    0   552    0     0    952      0 --:--:-- --:--:-- --:--:--   968
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<contextElementResponse>
  <contextElement>
    <entityId type="farm" isPattern="false">
      <id>farm1</id>
    </entityId>
    <contextAttributeList>
      <contextAttribute>
        <name>farm1_temp</name>
        <type>float</type>
        <contextValue>0</contextValue>
      </contextAttribute>
      <contextAttribute>
        <name>farm1_temp</name>
        <type>float</type>
        <contextValue>0</contextValue>
      </contextAttribute>
    </contextAttributeList>
    <domainMetadata/>
  </contextElement>
  <statusCode>
    <code>200</code>
    <reasonPhrase>OK</reasonPhrase>
  </statusCode>
</contextElementResponse>
```

I cannot get an temperature of a thermo1 and there are two farm1_temp attributes in a farm1 context...
Could you tell me how to use a NGSI-association feature in this case?

Best Regards,
Naoya

**Comment by Aeronbroker** [comment]
Dear Naoya,

the issues were multiple, both in the requests and in the IoT Broker code.

Following is the correct Association registration to be sent to the IoT Discovery.

- AssociationContext
```
(curl localhost:8060/ngsi9/registerContext -s -S \
    --header  'Content-Type: application/xml' --header 'Accept: application/xml' \
    -d @-) <<EOF
<?xml version="1.0" encoding="UTF-8"?>
<registerContextRequest>
<contextRegistrationList>
        <contextRegistration>
            <registrationMetadata>
                <contextMetadata>
                    <name>Association</name>
                    <type>Association</type>
                    <value>
                            <sourceEntityId type="thermometer" isPattern="false">
                                <id>thermo1</id>
                            </sourceEntityId>
                            <targetEntityId type="farm" isPattern="false">
                                <id>farm1</id>
                            </targetEntityId>
			<attributeAssociationList>
				<attributeAssociation>
        			<sourceAttribute>temperature</sourceAttribute>
       				<targetAttribute>farm1_temp</targetAttribute>
				</attributeAssociation>
			</attributeAssociationList>
                    </value>
                </contextMetadata>
            </registrationMetadata>
            <providingApplication>http://localhost:1027/ngsi10</providingApplication>
        </contextRegistration>
    </contextRegistrationList>
    <duration>P1Y</duration>
    <registrationId></registrationId>
</registerContextRequest>
EOF
```

As you can see the ContextRegistration for the Association needs to have 

- ContextMetadata.name set to "Association"
- The ContexteMetadata.value not wrapped in "entityAssociation"
- "attributeAssociationList" and "attributeAssociation" with the first letter as low-case.

Furthermore some bugs have been fixed in the code, so please make a git pull and try again your test.

Thank you very much for spotting this bug.

Best Regards
Flavio

**Comment by onlyindreams** [comment]
Dear Flavio,
Thank you for your kindness and great works!

In this case, I confirmed that a Association feature works fine but the feature seems to have some problems in other patterns yet ( for example, "Association of multiple attributes" or "Association of multiple entities" ).

Anyway, I decided to close this topic because I hate a too long topic.

Association problems I have yet will open as a new issue when I need your help.

Best Regards,
Naoya
